packages/costa/src/runtime.ts

```diff
@@ -0,0 +1,265 @@
+import Debug from 'debug';
+
+const debug = Debug('costa.runtime');
+const CONDA_CONFIG = path.join(__dirname, '../node_modules/@pipcook/boa/.CONDA_INSTALL_DIR');
+
+function selectLatestPackage(metadata: NpmPackageMetadata): NpmPackage {
+  return metadata.versions[metadata['dist-tags'].latest];
+}
+
+function spawnAsync(command: string, args?: string[], opts: SpawnOptions = {}): Promise<string> {
+  return new Promise((resolve, reject) => {
+    opts.stdio = 'inherit';
+    opts.detached = false;
+    const child = spawn(command, args, opts);
+    child.on('close', (code: number) => {
+      code === 0 ? resolve() : reject(new TypeError(`invalid code ${code} from ${command}`));
+    });
+  });
+}
+
+function createRequirements(name: string, config: CondaConfig): string[] {
+  const deps = [];
+  for (let k in config.dependencies) {
+    const v = config.dependencies[k];
+    if (v === '*') {
+      deps.push(k);
+    } else if (v.startsWith('git+https://') === true) {
+      deps.push(v);
+    } else {
+      deps.push(`${k}==${v}`);
+    }
+  }
+  return deps;
+}
+
+export { PluginPackage } from './index';
+export { RunnableResponse } from './runnable';
+export {
+  PluginRunnable,
+  BootstrapArg
+};
+
+/**
+ * The Costa runtime is for scheduling plugins and management.
+ */
+export class CostaRuntime {
+  options: RuntimeOptions;
+
+  /**
+   * Create a new Costa runtime by given `RuntimeOptions`.
+   * @param opts the runtime options.
+   */
+  constructor(opts: RuntimeOptions) {
+    this.options = opts;
+    process.env.NODE_PATH += `:${opts.installDir}/node_modules`;
+    ensureDirSync(opts.installDir);
+    ensureDirSync(opts.datasetDir);
+    ensureDirSync(opts.componentDir);
+  }
+  /**
+   * fetch and check if the package name is valid.
+   * @param name the plugin package name.
+   * @param cwd the current working directory.
+   */
+  async fetch(name: string, cwd?: string): Promise<PluginPackage> {
+    const source = this.getSource(name, cwd || process.cwd());
+    let pkg: PluginPackage;
+    if (source.from === 'npm') {
+      debug(`requesting the url ${source.uri}`);
+      const resp = await request(source.uri);
+      const meta = JSON.parse(resp) as NpmPackageMetadata;
+      pkg = selectLatestPackage(meta);
+    } else {
+      debug(`linking the url ${source.uri}`);
+      pkg = require(source.uri + '/package.json');
+    }
+
+    this.validPackage(pkg);
+    this.assignPackage(pkg, source);
+    return pkg;
+  }
+  /**
+   * Install the given plugin by name.
+   * @param name the plugin package name.
+   * @param force install from new anyway.
+   * @param pyIndex the index mirror to install python packages.
+   */
+  async install(pkg: PluginPackage, force = false, pyIndex?: string): Promise<boolean> {
+    // check if the pkg is installed
+    if ((await this.isInstalled(pkg.name)) && !force) {
+      debug(`skip install "${pkg.name}" because it already exists`);
+      return true;
+    }
```

Comment:
Does this operation conflict with npm's strategy? It also ignores the package version specified by the plugin, maybe this is an unnecessary step?

Reply:
NPM strategy is not transparent, and this step is required at Pipcook, so we need to make sure it works specifically.